src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs

```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
+                            Debug.Assert(!hProv.IsInvalid);
+                            Debug.Assert(!hProv.IsClosed);
+
+                            _safeProvHandle = hProv;
+                        }
+                    }
+
+                    return _safeProvHandle;
+                }
+
+                return _safeProvHandle;
+            }
+            set
             {
-                lock (this)
+                lock (_parameters)
                 {
-                    if (_safeKeyHandle == null)
+                    if (ReferenceEquals(value, _safeProvHandle))
                     {
-                        // We only attempt to generate a random key on desktop runtimes because the CoreCLR
-                        // RSA surface area is limited to simply verifying signatures.  Since generating a
-                        // random key to verify signatures will always lead to failure (unless we happend to
-                        // win the lottery and randomly generate the signing key ...), there is no need
-                        // to add this functionality to CoreCLR at this point.
-                        CapiHelper.GetKeyPairHelper(CapiHelper.CspAlgorithmType.Rsa, _parameters,
-                                            _randomKeyContainer, _keySize, ref _safeProvHandle, ref _safeKeyHandle);
+                        return;
                     }
+
+                    if (_safeProvHandle != null)
+                    {
+                        _safeKeyHandle?.Dispose();
+                        _safeKeyHandle = null;
+
+                        _safeProvHandle?.Dispose();
+                    }
+
+                    _safeProvHandle = value;
+                }
+            }
+        }
+
+        private SafeKeyHandle SafeKeyHandle
+        {
+            get
+            {
+                if (_safeKeyHandle == null)
+                {
+                    lock (_parameters)
+                    {
+                        if (_safeKeyHandle == null)
+                        {
+                            SafeKeyHandle hKey = CapiHelper.GetKeyPairHelper(
+                                CapiHelper.CspAlgorithmType.Rsa,
+                                _parameters,
+                                _keySize,
+                                SafeProvHandle);
+
+                            Debug.Assert(hKey != null);
+                            Debug.Assert(!hKey.IsInvalid);
+                            Debug.Assert(!hKey.IsClosed);
+
+                            _safeKeyHandle = hKey;
+                        }
+                    }
+                }
+
+                return _safeKeyHandle;
+            }
+
+            set
+            {
+                lock (_parameters)
+                {
+                    if (ReferenceEquals(value, _safeKeyHandle))
+                    {
+                        return;
+                    }
+
+                    _safeKeyHandle?.Dispose();
```

Comment:
Is
```C#
set
{
    lock (_parameters)
    {
        SafeKeyHandle current = _safeKeyHandle;

        if (ReferenceEquals(value, current))
        {
            return;
        }

        _safeKeyHandle = value;
        current?.Dispose();
    }
}
```

sufficient?  The get can still have read the old field, but that's also possible with any of the dirty reads that the code was already doing.  It does eliminate the dirty read in the get seeing the disposed value right between dispose and reassignment.

And if we feel more comfortable just unrolling this back to what it used to be, that's fine.  This just seemed as correct and more robust to me.  (As an example, importing a public-only blob overwrote _safeProvHandle, but didn't dispose out the previous one, and it let the stale key handle float for another couple of lines)